src/map/instance.cpp

```diff
@@ -66,240 +82,251 @@ struct instance_db *instance_searchname_db(const char *instance_name) {
 	}
 	return;
 }
 
-/*==========================================
+/**
  * Deletes an instance timer (Destroys instance)
- *------------------------------------------*/
+ */
 static TIMER_FUNC(instance_delete_timer){
 	instance_destroy(id);
 
 	return 0;
 }
 
-/*==========================================
+/**
  * Create subscription timer
- *------------------------------------------*/
+ */
 static TIMER_FUNC(instance_subscription_timer){
-	int i, ret;
-	unsigned short instance_id = instance_wait.id[0];
+	int ret;
+	uint16 instance_id = instance_wait.id[0];
 	struct map_session_data *sd = NULL;
 	struct party_data *pd = NULL;
 	struct guild *gd = NULL;
 	struct clan *cd = NULL;
-	enum instance_mode mode;
+	enum e_instance_mode mode;
 
-	if(instance_wait.count == 0 || instance_id == 0)
+	if (instance_wait.id.size() == 0 || instance_id == 0)
 		return 0;
 
 	// Check that maps have been added
 	ret = instance_addmap(instance_id);
-	mode = instance_data[instance_id].mode;
+
+	auto idata = instance_search(instance_id);
+
+	mode = idata->mode;
 
 	switch(mode) {
 		case IM_NONE:
 			break;
 		case IM_CHAR:
-			if (ret == 0 && (sd = map_charid2sd(instance_data[instance_id].owner_id)) != NULL) // If no maps are created, tell player to wait
+			if (ret == 0 && (sd = map_charid2sd(idata->owner_id)) != NULL) // If no maps are created, tell player to wait
 				clif_instance_changewait(instance_id, 0xffff);
 			break;
 		case IM_PARTY:
-			if (ret == 0 && (pd = party_search(instance_data[instance_id].owner_id)) != NULL) // If no maps are created, tell party to wait
+			if (ret == 0 && (pd = party_search(idata->owner_id)) != NULL) // If no maps are created, tell party to wait
 				clif_instance_changewait(instance_id, 0xffff);
 			break;
 		case IM_GUILD:
-			if (ret == 0 && (gd = guild_search(instance_data[instance_id].owner_id)) != NULL) // If no maps are created, tell guild to wait
+			if (ret == 0 && (gd = guild_search(idata->owner_id)) != NULL) // If no maps are created, tell guild to wait
 				clif_instance_changewait(instance_id, 0xffff);
 			break;
 		case IM_CLAN:
-			if (ret == 0 && (cd = clan_search(instance_data[instance_id].owner_id)) != NULL) // If no maps are created, tell clan to wait
+			if (ret == 0 && (cd = clan_search(idata->owner_id)) != NULL) // If no maps are created, tell clan to wait
 				clif_instance_changewait(instance_id, 0xffff);
 			break;
 		default:
 			return 0;
 	}
 
-	instance_wait.count--;
-	memmove(&instance_wait.id[0],&instance_wait.id[1],sizeof(instance_wait.id[0])*instance_wait.count);
-	memset(&instance_wait.id[instance_wait.count], 0, sizeof(instance_wait.id[0]));
+	instance_wait.id.pop_front();
 
-	for(i = 0; i < instance_wait.count; i++) {
-		if(	instance_data[instance_wait.id[i]].state == INSTANCE_IDLE &&
-			((mode == IM_CHAR && sd != NULL) || (mode == IM_GUILD && gd != NULL) || (mode == IM_PARTY && pd != NULL) || (mode == IM_CLAN && cd != NULL))
-		){
+	for(uint16 i = 0; i < instance_wait.id.size(); i++) {
```

Comment:
on debug build the fonction size will be called n time. might be best to move it out the loop.